### PR TITLE
add callback to linear_svm

### PR DESCRIPTION
**Pull request by jaibhageria**
Addressing issue #2013  to add callback parameters to linear_svm

### CODE REVIEW

File: src/mlpack/tests/callback_test.cpp [hunk 1]
```diff
@@ -115,4 +118,27 @@ BOOST_AUTO_TEST_CASE(RNNWithOptimizerCallbackTest)
...
+  arma::mat dataset = "2 0 0;"
+                      "0 0 0;"
+                      "0 2 1;"
+                      "1 0 2;"
+                      "0 1 0";
+
+  //  Corresponding labels
+  arma::Row<size_t> labels = "1 0 1";
+
+  LinearSVMFunction<arma::mat> svm(dataset, labels, 2, 0.0);
```
Comment (Yashwants19) [inline src/mlpack/tests/callback_test.cpp]: Hi @jaibhageria You can use default constructor here.

File: src/mlpack/tests/callback_test.cpp [hunk 2]
```diff
@@ -115,4 +118,27 @@ BOOST_AUTO_TEST_CASE(RNNWithOptimizerCallbackTest)
...
+
+  std::stringstream stream;
+  svm.Train(ens::PrintLoss(stream));
```
Comment (Yashwants19) [inline src/mlpack/tests/callback_test.cpp]: I think so passing `dataset` ,`labels` and `numClasses` here will work.

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 3]
```diff
@@ -204,11 +204,12 @@ class LinearSVM
    * @param optimizer Desired optimizer.
    * @return Objective value of the final point.
```
Comment (Yashwants19) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Adding documentation for callbacks will be great.

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 4]
```diff
@@ -202,13 +202,15 @@ class LinearSVM
...
    */
-  template <typename OptimizerType = ens::L_BFGS>
+  template <typename OptimizerType = ens::L_BFGS, typename... CallbackTypes>
   double Train(const MatType& data,
                const arma::Row<size_t>& labels,
                const size_t numClasses = 2,
-               OptimizerType optimizer = OptimizerType());
+               OptimizerType optimizer = OptimizerType(),
+               CallbackTypes&&... callbacks);
```
Comment (zoq) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Can you add a description for the template parameter as well? See https://github.com/mlpack/mlpack/blob/db03fbf4838a4e6226f0f332857528ab195bf4ec/src/mlpack/methods/linear_svm/linear_svm.hpp#L199 for en example.

**Comment by jaibhageria** [comment]
@zoq I am getting Callbacktest failure in Travis as follows
```
        Start  16: CallbackTest
4498 13/155 Test  #16: CallbackTest .....................***Failed    0.42 sec
4499Running 5 test cases...
4500/home/travis/build/mlpack/mlpack/src/mlpack/tests/callback_test.cpp(142): fatal error in
"LinearSVMCallbackTest": critical check stream.str().length() > 0 failed [0 <= 0]
```
Not able to figure out why PrintLoss stream has size>0. 
Can you please help?

**Comment by vss96** [comment]
Hey @jaibhageria, you have to change your optimizer to make it work (BFGS does not work with what you are trying to test). Try using StandardSGG or something similar. You can look at the other tests to make sense out of it.

**Comment by jaibhageria** [comment]
@zoq I have tried changing optimiser function to Standard SGD and it resulted in failure. 
Now I have changed optimiser to Gradient Descent by referring to the following lines of code as can be seen : 
https://github.com/mlpack/mlpack/blob/8da09477766a4c1e65b38e8de4155e52213e73ba/src/mlpack/tests/linear_svm_test.cpp#L500-L502

But I still end up getting this error in the Travis builds : 
```
13/155 Test  #16: CallbackTest .....................***Failed    0.43 sec
4513Running 5 test cases...
4514/home/travis/build/mlpack/mlpack/src/mlpack/tests/callback_test.cpp(151): fatal error in
"LinearSVMCallbackTest": critical check stream.str().length() > 0 failed [0 <= 0]
```

Please suggest the required code changes.

**Comment by vss96** [comment]
@jaibhageria There are a few other optimizers which you can try out. If you have the build setup on your local, you can check if the tests are passing by doing a "ctest -j2". It will help you debug faster rather than push every time and then check whether it works or not.

**Comment by zoq** [comment]
You can use:

```c++
class CallbackTestFunction
{
 public:
    CallbackTestFunction() : calledEndOptimization(false) {}

    template<typename OptimizerType, typename FunctionType, typename MatType>
    void EndOptimization(OptimizerType& /* optimizer */,
                       FunctionType& /* function */,
                       MatType& /* coordinates */)
    {
        calledEndOptimization = true;
    }

    bool calledEndOptimization;
}
```

instead of the `ProgressBar` and check if `calledEndOptimization` is true at the end.

**Comment by zoq** [comment]
> @jaibhageria There are a few other optimizers which you can try out. If you have the build setup on your local, you can check if the tests are passing by doing a "ctest -j2". It will help you debug faster rather than push every time and then check whether it works or not.

You can even run a single test case using: `bin/mlpack_test -t TestSuite/TestCase` e.g.   `bin/mlpack_test -t CallbackTest/LinearSVMCallbackTest`  to run the `LinearSVMCallbackTest` test from the `CallbackTest` suite.

**Comment by jaibhageria** [comment]
Sorry for the delay, as I got a bit busy. I will be working to fix this issue soon

**Comment by zoq** [comment]
No worries, always challenging to find the time.

**Comment by Yashwants19** [comment]
Hi @jaibhageria It will be great if you can add callbacks in constructor too where we call train function.

File: src/mlpack/tests/callback_test.cpp [hunk 5]
```diff
@@ -135,4 +138,34 @@ BOOST_AUTO_TEST_CASE(LRWithOptimizerCallback)
...
+  //  Corresponding labels
+  arma::Row<size_t> labels = "1 0 1";
+
+  CallbackTestFunction cb;
+
+  ens::L_BFGS opt;
+  LinearSVM<arma::mat> lsvm(dataset, labels, numClasses, lambda,
+      delta, false, opt, cb);
+
+  lsvm.Train(dataset, labels, numClasses, opt, cb);
```
Comment (Yashwants19) [inline src/mlpack/tests/callback_test.cpp]: I think so you can remove this because train function is also calling in above constructor.

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 6]
```diff
@@ -94,14 +94,15 @@ class LinearSVM
...
-            const double lambda = 0.0001,
-            const double delta = 1.0,
-            const bool fitIntercept = false,
-            OptimizerType optimizer = OptimizerType());
+            const size_t numClasses,
+            const double lambda,
+            const double delta,
+            const bool fitIntercept,
+            OptimizerType optimizer,
+            CallbackTypes&&... callbacks);
```
Comment (Yashwants19) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Can you please add description for callbacks in this constructor and now this is not default we should remove that part also.

Comment (jaibhageria) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Done.

File: src/mlpack/tests/callback_test.cpp [hunk 7]
```diff
@@ -135,4 +138,32 @@ BOOST_AUTO_TEST_CASE(LRWithOptimizerCallback)
...
+/**
+ * Test a Linear SVM model with EndOptimization callback.
+ */
+BOOST_AUTO_TEST_CASE(LinearSVMCallbackTest)
+{
+  const size_t numClasses = 2;
+  const double lambda = 0.0001;
+  const double delta = 1.0;
+
+  // A very simple fake dataset
```
Comment (zoq) [inline src/mlpack/tests/callback_test.cpp]: Pedantic style issue, missing stop at the end.

File: src/mlpack/tests/callback_test.cpp [hunk 8]
```diff
@@ -135,4 +138,32 @@ BOOST_AUTO_TEST_CASE(LRWithOptimizerCallback)
...
+  arma::mat dataset = "2 0 0;"
+                      "0 0 0;"
+                      "0 2 1;"
+                      "1 0 2;"
+                      "0 1 0";
+
+  //  Corresponding labels
```
Comment (zoq) [inline src/mlpack/tests/callback_test.cpp]: See comment above, also looks like you used two spaces after `//`, but only one is needed; sometimes I'm very picky about the style.

File: src/mlpack/tests/linear_svm_test.cpp [hunk 9]
```diff
@@ -543,8 +548,11 @@ BOOST_AUTO_TEST_CASE(LinearSVMLBFGSTwoClasses)
       labels(i) = 1;
     }
 
+    CallbackTestFunction cb;
+    ens::L_BFGS opt;
     // Create a linear svm object using L-BFGS optimizer.
-    LinearSVM<arma::mat> lsvm(data, labels, numClasses, lambda);
+    LinearSVM<arma::mat> lsvm(data, labels, numClasses, lambda,
```
Comment (zoq) [inline src/mlpack/tests/linear_svm_test.cpp]: I think it's fine to test the model without any callbacks, as it should work as well and should also be tested, so do you mind to revert the changes here?

**Comment by jaibhageria** [comment]
@zoq @Yashwants19 
I am getting a Travis Build timeout error for Travis build #8236.1
```
Still running (7 of 30): ctest -j2
6212
6213The job exceeded the maximum time limit for jobs, and has been terminated.
```

I have made the required code changes and tested everything locally, which seems to work perfectly. Please let me know if there any other changes to be made.

**Comment by Yashwants19** [comment]
This looks good to me.

**Comment by zoq** [comment]
> @zoq @Yashwants19
> I am getting a Travis Build timeout error for Travis build #8236.1
> 
> ```
> Still running (7 of 30): ctest -j2
> 6212
> 6213The job exceeded the maximum time limit for jobs, and has been terminated.
> ```
> 
> I have made the required code changes and tested everything locally, which seems to work perfectly. Please let me know if there any other changes to be made.

Once https://github.com/mlpack/mlpack/pull/2058 is ready and merged it should solve the build timeout, at this point we don't have to worry about the timeout since the other builds are passing.

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 10]
```diff
@@ -264,6 +270,26 @@ class LinearSVM
...
+
+    template<typename OptimizerType, typename FunctionType, typename MatType>
+    void EndOptimization(OptimizerType& /* optimizer */,
+                       FunctionType& /* function */,
+                       MatType& /* coordinates */)
+    {
+        calledEndOptimization = true;
+    }
+
+    //! Track to check if callback is executed
```
Comment (zoq) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Sorry another picky style comment, there is a missing stop at the end.

Comment (jaibhageria) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Done.

**Comment by jaibhageria** [comment]
@zoq Thank you! I guess I have made all the required changes.

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 11]
```diff
@@ -264,6 +270,26 @@ class LinearSVM
...
+    void EndOptimization(OptimizerType& /* optimizer */,
+                       FunctionType& /* function */,
+                       MatType& /* coordinates */)
+    {
+        calledEndOptimization = true;
+    }
+
+    //! Track to check if callback is executed.
+    bool calledEndOptimization;
+};
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Can we put this in the test code directory?

File: src/mlpack/tests/linear_svm_test.cpp [hunk 12]
```diff
@@ -454,19 +454,22 @@ BOOST_AUTO_TEST_CASE(LinearSVMLBFGSSimpleTest)
...
 
-  //  Corresponding labels
+  // Corresponding labels.
   arma::Row<size_t> labels = "1 0 1";
 
+  ens::L_BFGS opt;
   // Create a linear svm object using L-BFGS optimizer.
-  LinearSVM<arma::mat> lsvm(dataset, labels, numClasses, lambda);
+  LinearSVM<arma::mat> lsvm(dataset, labels, numClasses, lambda,
+    delta, false, opt);
```
Comment (rcurtin) [inline src/mlpack/tests/linear_svm_test.cpp]: Oops, minor style issue---can you indent the wrapped line twice?  (This happens a few times)

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 13]
```diff
@@ -82,26 +82,29 @@ class LinearSVM
...
-            const size_t numClasses = 2,
-            const double lambda = 0.0001,
-            const double delta = 1.0,
-            const bool fitIntercept = false,
-            OptimizerType optimizer = OptimizerType());
+            const size_t numClasses,
+            const double lambda,
+            const double delta,
+            const bool fitIntercept,
+            OptimizerType optimizer,
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: With this change there's now no way to avoid specifying an optimizer.  Do you think we can re-add the `optimizer` default?  This should be allowed even if `callbacks` does not have a default:

https://stackoverflow.com/questions/29098835/can-parameter-pack-function-arguments-be-defaulted

Alternately, if you have an issue making that work, you could add another overload of `Train()` and the constructor (so that one overload has callbacks, and the other doesn't).

Comment (zoq) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Nice catch!

**Comment by rcurtin** [comment]
Looks like I have an mlpack-bot bug to track down :(

**Comment by jaibhageria** [comment]
@rcurtin I will surely be working to resolve the above issues. Please give me some time as I would be busy for the next couple of days.

File: src/mlpack/methods/linear_svm/linear_svm.hpp (hunk 11 above)
Comment (jaibhageria) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: @rcurtin Should I create a new file as "linear_svm_test_tools.hpp" and place the           ```class CallbackTestFunction``` under it ?

File: src/mlpack/methods/linear_svm/linear_svm.hpp (hunk 13 above)
Comment (jaibhageria) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: @rcurtin I will have to do an overload for ```Train()``` and ```LinearSVM``` constructor.
I have tried the method suggested in the link but it throws errors.

File: src/mlpack/methods/linear_svm/linear_svm.hpp (hunk 11 above)
Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: I think it would be fine to just put the class definition in `linear_svm_test.cpp`.

File: src/mlpack/methods/linear_svm/linear_svm.hpp (hunk 13 above)
Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Yeah, the Stack Overflow post suggests that not every compiler will handle that correctly.  So multiple overloads is fine. :+1:

File: src/mlpack/methods/linear_svm/linear_svm.hpp (hunk 11 above)
Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Test comment for mlpack-bot...

Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Test comment for mlpack-bot...

Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Test comment for mlpack-bot...

Comment (rcurtin) [reply src/mlpack/methods/linear_svm/linear_svm.hpp]: Test comment for mlpack-bot...

**Comment by rcurtin** [comment]
Final test comment for mlpack-bot... (I think I have it solved.  Hopefully mlpack-bot will not respond offering more stickers 😄).

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 14]
```diff
@@ -196,19 +199,22 @@ class LinearSVM
...
-   * @tparam OptimizerType Desired optimizer
-   * @param data Input training features. Each column associate with one sample
+   * @tparam OptimizerType Desired optimizer.
+   * @tparam CallbackTypes Types of Callback Functions.
+   * @param data Input training features. Each column associate with one sample.
    * @param labels Labels associated with the feature data.
    * @param numClasses Number of classes for classification.
    * @param lambda L2-regularization constant.
    * @param optimizer Desired optimizer.
+   * @param callbacks Callback Functions.
```
Comment (zoq) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: In another PR, @rcurtin suggested to add a reference to the ensmallen documentation: https://github.com/mlpack/mlpack/pull/2074#discussion_r341866123. I think we should do the same here as well.

**Comment by jaibhageria** [comment]
Hey, @rcurtin and @zoq I have made the required changes. Can you please review and merge?

**Comment by rcurtin** [comment]
Hi @jaibhageria, there are a lot of PRs and not enough time.  I'll try to review this when I have a chance. :+1:

File: src/mlpack/methods/linear_svm/linear_svm.hpp [hunk 15]
```diff
@@ -196,11 +225,30 @@ class LinearSVM
...
+               CallbackTypes&&... callbacks);
+
+  /**
+   * Train the Linear SVM with the given training data.
+   *
+   * @tparam OptimizerType Desired optimizer.
+   * @param data Input training features. Each column associate with one sample.
    * @param labels Labels associated with the feature data.
    * @param numClasses Number of classes for classification.
-   * @param lambda L2-regularization constant.
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm.hpp]: Nice catch, thanks! :+1:

File: src/mlpack/methods/linear_svm/linear_svm_impl.hpp [hunk 16]
```diff
@@ -66,6 +85,36 @@ LinearSVM<MatType>::LinearSVM(
...
+  // Train the model.
+  Timer::Start("linear_svm_optimization");
+  const double out = optimizer.Optimize(svm, parameters, callbacks...);
+  Timer::Stop("linear_svm_optimization");
+
+  Log::Info << "LinearSVM::LinearSVM(): final objective of "
+            << "trained model is " << out << "." << std::endl;
+
+  return out;
+}
```
Comment (rcurtin) [inline src/mlpack/methods/linear_svm/linear_svm_impl.hpp]: For a minute, I thought it would be nice to replace the implementation of the `Train()` overload that doesn't take callbacks with a call to this overload (with nothing as the callbacks).  But... I don't know how to call this overload from that one; I'm not sure it's possible.  So maybe the duplication can't be avoided.  If you know some way, let me know. :)

File: src/mlpack/tests/callback_test.cpp [hunk 17]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (rcurtin) [inline src/mlpack/tests/callback_test.cpp]: Thanks, I really appreciate small fixes like this.  For this one though we should make sure to add a comment that, among other things, specifies the license, like this:

```
/**
 * @file callback_test.cpp
 *
 * Tests for ensmallen callback functionality.
 *
 * mlpack is free software; you may redistribute it and/or modify it under the
 * terms of the 3-clause BSD license.  You should have received a copy of the
 * 3-clause BSD license along with mlpack.  If not, see
 * http://www.opensource.org/licenses/BSD-3-Clause for more information.
 */
```

If you want, I can handle it during merge. :+1:

Comment (rcurtin) [reply src/mlpack/tests/callback_test.cpp]: Handled with 90014a6f9.

**Comment by rcurtin** [comment]
Thanks for the contribution! :+1:
